Create specialized macros for each operator group

DarkDrek opened the pull request:
Fix mixed type checking for binops that were forcing first argument type as second now the expected type always depends on the operator
Generalize type_check fail tests

New tests for #20
- all int comparison operators with bool arguments
- all equality operators with (int, bool) and (bool, int) arguments
- all logical operators with (int, bool) and (bool, int) arguments

pvdrz commented on src/ty/ty_check.rs:
```diff
@@ -16,6 +16,10 @@ pub fn expect_ty(expected: Ty, found: Ty) -> TyResult<Ty> {
     }
 }
 
+fn expect_ty2(expected: Ty, ty1: Ty, ty2: Ty) -> TyResult<Ty> {
```
Hmmm This makes me think we might as well write a macro for this. Something like `ensure_ty!(expected, ty1, ty2, ...)`

DarkDrek replied:
Or if performance isn't a major concern a iterator based method would also work. But since I needed just one more checked parameter I didn't generalized it further. If we add more complex (ternary operators anyone?) checks we can still decide then.

pvdrz replied:
my only complain with an iterator based method is the hassle of wrapping the types you want to check in an iterable structure such as a vector or a slice:
```rust
ensure_ty(Ty::Bool, vec![ty1, ty2])?;
```
With a single type is even worse because you end up with
```rust
ensure_ty(Ty::Bool, vec![ty])?;
```
So we'd have to keep the old `expect_ty` for one type and have `ensure_ty` for several. Which isn't great.  I think that the macro is an easy way to have this variable argument function without bothering with wrapping your arguments to do so.

DarkDrek replied:
Another thing is that it is possible to provide an empty iterator/slice. Should I (try to) add the macro now?

pvdrz replied:
it would be cool. For now it's enough to just match two or three comma separated arguments.

pvdrz commented on tests/type_check/mod.rs:
```diff
@@ -48,3 +48,59 @@ macro_rules! test_type_with_placeholder {
+/// in the same directory against the `$pattern`. It does this once for
+/// each int binary operator that will replace the `$placeholder` in the file.
+#[macro_export]
+macro_rules! test_type_with_all_int_binops {
```
I think that `test_type_for_all_integer_binops` sounds better. Same for the other macros.

DarkDrek committed:
Create specialized macros for each operator group

Fix mixed type checking for binops forcing first argument type as second
Generalize type_check fail test
Rename macros

DarkDrek committed:
Replace expect_ty2 with ensure_ty macro
